File: Euclid/Messages/Outgoing/STATUS.cs
using Euclid.Game;
using Euclid.Game.Entity.Humanoid;
using Euclid.Network.Streams.Util;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Messages.Outgoing
{
    class STATUS : IMessageComposer
    {
        public EntityState[] entities;

        public STATUS(ICollection<Humanoid> values)
        {
            this.entities = values.Select(x => new
                EntityState(
                x.EntityData.Id,
                x.RoomEntity.InstanceId,
                x.EntityData,
                x.EntityType,
                x.RoomEntity.Room,
                x.RoomEntity.Position.Copy(),
                x.RoomEntity.Status)
             ).ToArray();
        }

        public override void Write()
        {
            foreach (var entity in this.entities)
            {
                Data.Add(new ArgumentEntry(entity.Details.Name));
                Data.Add(new DelimeterEntry(entity.Position.X, ' '));
                Data.Add(new DelimeterEntry(entity.Position.Y, ','));
                Data.Add(new DelimeterEntry((int)entity.Position.Z, ','));
                Data.Add(new DelimeterEntry(entity.Position.HeadRotation, ','));
                Data.Add(new DelimeterEntry(entity.Position.BodyRotation, ','));
                Data.Add("/");

                foreach (var kvp in entity.Statuses)
                {
                    Data.Add(kvp.Key);

                    if (kvp.Value.Value.Length > 0)
                    {
                        Data.Add(" ");
                        Data.Add(kvp.Value.Value);
                    }

                    Data.Add("/");
                }
            }
        }
    }
}

File: euclid/src/main/java/euclid/messages/outgoing/STATUS.java
package euclid.messages.outgoing;

import euclid.game.entity.EntityState;
import euclid.game.entity.Humanoid;
import euclid.messages.IMessageComposer;
import euclid.network.streams.util.ArgumentEntry;
import euclid.network.streams.util.DelimeterEntry;

import java.util.Collection;

public class STATUS extends IMessageComposer {

    private final EntityState[] entities;

    public STATUS(Collection<Humanoid> values) {
        this.entities = values.stream()
                .map(x -> new EntityState(
                        x.getEntityData().getId(),
                        x.getRoomEntity().getInstanceId(),
                        x.getEntityData(),
                        x.getEntityType(),
                        x.getRoomEntity().getRoom(),
                        x.getRoomEntity().getPosition().copy(),
                        x.getRoomEntity().getStatuses()
                ))
                .toArray(EntityState[]::new);
    }

    @Override
    public void write() {
        for (EntityState entity : entities) {
            getData().add(new ArgumentEntry(entity.getDetails().getName()));
            getData().add(new DelimeterEntry(entity.getPosition().getX(), " "));
            getData().add(new DelimeterEntry(entity.getPosition().getY(), ","));
            getData().add(new DelimeterEntry((int) entity.getPosition().getZ(), ","));
            getData().add(new DelimeterEntry(entity.getPosition().getHeadRotation(), ","));
            getData().add(new DelimeterEntry(entity.getPosition().getBodyRotation(), ","));
            getData().add("/");

            entity.getStatuses().forEach((key, status) -> {
                getData().add(key);
                if (status.getValue().length() > 0) {
                    getData().add(" ");
                    getData().add(status.getValue());
                }
                getData().add("/");
            });
        }
    }
}
